Add unified structured logging middleware for Rack applications

corrupt952 opened the pull request:
## Summary

Implement a comprehensive Rack middleware for unified structured logging following Studist's common log format. This middleware provides standardized JSON/LTSV logging with 18 predefined fields for consistent log analysis across all services.

## Key Features

  - **Structured Logging**: JSON and LTSV format support with 18 standardized fields
  - **Configurable Extractors**: Custom logic for user_id, user_group_id, user_authority, and normalized_uri
  - **AWS X-Ray Compatible**: Built-in trace_id extraction for distributed tracing
  - **Zero Configuration**: Works out of the box with sensible defaults
  - **Performance Focused**: Minimal overhead with efficient field extraction
  - **Error Handling**: Proper exception logging with 500 status codes

## Implementation Details

### Core Architecture
  - `Studist::Rack::Logger`: Main entry point and factory method
  - `Middleware`: Core request/response processing and timing measurement
  - `LogEntryBuilder`: Extracts and organizes 18 standardized log fields
  - `Formatter::Json/Ltsv`: Output formatters for different log formats

### Log Fields (18 total)
  - **Timing**: timestamp, response_time_ms
  - **Request**: method, URL, headers, body_size, query_string
  - **Response**: status_code, body_size
  - **Infrastructure**: server_name, trace_id, request_id, forwarded headers
  - **User Context**: user_id, user_group_id, user_authority (via extractors)
  - **Routing**: normalized_uri (via extractor)

### Quality Assurance
  - **100% YARD Documentation**: Comprehensive API documentation for all public methods
  - **Comprehensive Testing**: RSpec tests with Ruby 3.0-3.4 and Rack 2.0-3.1 matrix
  - **Code Quality**: RuboCop linting with consistent style enforcement
  - **CI/CD Ready**: GitHub Actions workflow for automated testing and validation

## Usage Examples

  ```ruby
  # Basic usage
  use Studist::Rack::Logger, app_id: 'my-service'

  # Advanced configuration
  use Studist::Rack::Logger,
    app_id: 'my-service',
    format: :json,
    logger: Rails.logger,
    user_id_extractor: ->(env, req) { env['user.id'] },
    normalized_uri_extractor: ->(env, req) { normalize_path(req.path) }
  ```

## Testing

- ✅ Ruby 3.0.7, 3.1.6, 3.2.6, 3.3.7, 3.4.3
- ✅ Rack 2.0.9, 2.2.9, 3.0.12, 3.1.15
- ✅ RuboCop style compliance
- ✅ 100% YARD documentation coverage
- ✅ Comprehensive unit and integration tests

corrupt952 committed:
chore: Configure Rubocop and RSpec

Setup tools based on existing repo setup

corrupt952 committed:
feat: implement unified logging middleware for Rack applications

Add comprehensive studist-rack-logger gem with structured JSON/LTSV logging supporting 18 standardized fields including timestamp, trace_id, user information, and response  metrics.

corrupt952 committed:
chore: add .rspec_status to .gitignore

Exclude RSpec status persistence file from version control

corrupt952 committed:
docs: Add comprehensive YARD documentation for API

- Add YARD development dependency to gemspec for documentation generation
- Create .yardopts configuration file for consistent doc generation
- Add comprehensive API documentation to all public classes and methods
- Document namespace modules (Studist, Studist::Rack) with clear descriptions
- Add detailed examples and parameter documentation for middleware usage
- Achieve 100% documentation coverage for enhanced developer experience
- Enable RubyGems.org auto-documentation and IDE IntelliSense support

corrupt952 committed:
chore: Update Ruby version requirement to 3.0+

- Change required_ruby_version from '>= 2.5.9, <= 3.4.3' to '>= 3.0.0, < 3.5.0'
- Drop support for EOL Ruby 2.x versions (security and maintenance)
- Align with modern Ruby ecosystem and best practices
- Support Ruby 3.0 through 3.4 for current and future compatibility

corrupt952 committed:
ci: Add comprehensive GitHub Actions workflow for testing

- Add CI workflow testing Ruby 3.0.7, 3.1.6, 3.2.6, 3.3.7, 3.4.3
- Test against Rack versions 2.0.9, 2.2.9, 3.0.12, 3.1.15 in matrix
- Include RuboCop linting and YARD documentation coverage checks
- Configure proper bundler caching with version-specific cache keys
- Trigger on pull requests to main branch for efficient CI execution
- Add gem build validation to ensure release readiness\

corrupt952 committed:
chore: rubocop -a

corrupt952 committed:
chore: organize development gems into proper group

corrupt952 committed:
refactor: remove unused methods and reorganize formatter structure

corrupt952 committed:
refactor: replace def app with let(:app) in specs to avoid global method pollution

corrupt952 committed:
feat: add simple configuration DSL and performance optimizations

- Add Configuration class with DSL support for better option management
- Implement performance optimizations (hostname caching, sampling, request filtering)
- Enhance error handling with structured logging and safe fallbacks
- Add comprehensive filtering system (skip_paths, skip_if conditions)
- Support sampling rates for both normal requests and errors
- Maintain backward compatibility with hash-based configuration
- Add extensive test coverage for new configuration features

corrupt952 committed:
refactor: replace response hash with Struct for type safety

Introduce Response Struct to standardize response data handling,
replacing direct parameter passing with structured objects.

corrupt952 committed:
test: add comprehensive log output verification for all 18 specification fields

fumi23 commented on lib/studist/rack/logger/log_entry_builder.rb:
```diff
@@ -0,0 +1,156 @@
+          def extract_remote_addr(env)
+            forwarded_for = env['HTTP_X_FORWARDED_FOR']
+            if forwarded_for
+              forwarded_for.split(',').first.strip
```
trusted proxyを除外する処理はいりませんか？

corrupt952 replied:
ありがとうございます。
差分がクソでかくて申し訳ないですが、一応実装しておきました。
https://github.com/StudistCorporation/studist-rack-logger/pull/6/commits/cea4456f488bd214530ae48d4c6d9ca85c9c457e

fumi23 replied:
このミドルウェアのconfigとして実装されたと思うんですが、
これってRailsのconfigを反映させられないですか （・ω・`）？

corrupt952 replied:
> これってRailsのconfigを反映させられないですか （・ω・`）？

Rails用のMiddlewareではなく、**Rackの**Middlewareとして実装しているので、あまり対応はしたくないですね。
仮にやるとしても別Issueたてて対応とかにするんじゃないですかね。

fumi23 replied:
すごい差分だった https://github.com/StudistCorporation/studist-rack-logger/pull/6/commits/cea4456f488bd214530ae48d4c6d9ca85c9c457e （・ω・`）

気軽に言ったけど、まじめにつくるとこんなに大変なのか

fumi23 replied:
> Rails用のMiddlewareではなく、RackのMiddlewareとして実装しているので、あまり対応はしたくない

理解 🙆

fumi23 commented on lib/studist/rack/logger/middleware.rb:
```diff
@@ -0,0 +1,203 @@
+            return [] unless backtrace
+
+            # Limit backtrace size and filter out gem paths for cleaner logs
+            backtrace.take(10).reject { |line| line.include?('/gems/') }
```
うれしい

fumi23 reviewed (commented):
よみました◎

corrupt952 committed:
feat: implement trusted proxy filtering for secure IP extraction

- Add RemoteIp class with Rails-compatible trusted proxy logic
- Support configurable trusted proxy ranges with RFC 1918 defaults
- Implement secure header processing priority (X-Forwarded-For → X-Real-IP →
Client-IP → REMOTE_ADDR)
- Handle IPv4-mapped IPv6 addresses for Ruby 3.1+ compatibility
- Add comprehensive edge case handling (empty headers, malformed chains, invalid IPs)
- Include performance tests for large IP chains and concurrent access
- Integrate trusted proxy filtering into log entry builder
- Update middleware tests to verify trusted proxy behavior
- Protect against IP spoofing from untrusted proxies
- Validate all IP addresses before processing
- Filter private network ranges automatically
- Graceful fallback for edge cases

corrupt952 committed:
fix: correct trusted proxy filtering and reduce code complexity

Fixes IP extraction logic to properly validate trusted proxies and refactors
complex methods into smaller, focused functions for better maintainability.

corrupt952 committed:
chore: remove performance spec

fumi23 reviewed (approved):
みました◎
